velox/exec/fuzzer/WindowFuzzer.cpp

```diff
@@ -68,20 +68,66 @@ void WindowFuzzer::addWindowFunctionSignatures(
   }
 }
 
-std::tuple<std::string, bool> WindowFuzzer::generateFrameClause() {
-  auto frameType = [](int value) -> const std::string {
+void WindowFuzzer::addKRangeFrameColumnToInput(
+    std::vector<RowVectorPtr>& input,
+    core::WindowNode::BoundType frameBoundType,
+    const std::string& columnName) {
+  // For frames with k RANGE PRECEDING/FOLLOWING, Velox requires the application
+  // to add columns with the range frame boundary value computed according
+  // to the frame type.
+  // If the frame is k PRECEDING :
+  // frame_boundary_value = current_order_by - k (for ascending ORDER BY)
+  // frame_boundary_value = current_order_by + k (for descending ORDER BY)
+  // If the frame is k FOLLOWING :
+  // frame_boundary_value = current_order_by + k (for ascending ORDER BY)
+  // frame_boundary_value = current_order_by - k (for descending ORDER BY)
+  auto computeOffset = [&](const core::WindowNode::BoundType& frameBoundType,
+                           const int64_t& offsetValue,
+                           const bool& ascending) -> int64_t {
+    if (frameBoundType == core::WindowNode::BoundType::kPreceding) {
+      return ascending ? -offsetValue : offsetValue;
+    } else {
+      return ascending ? offsetValue : -offsetValue;
+    }
+  };
+
+  auto size = vectorFuzzer_.getOptions().vectorSize;
+  velox::test::VectorMaker vectorMaker{pool_.get()};
+
+  int64_t offsetValue =
+      boost::random::uniform_int_distribution<int64_t>(1, 10)(rng_);
+  // Setting ascending boolean to true since we're using ascending row_number
+  // column as the order by column
+  int64_t offset = computeOffset(frameBoundType, offsetValue, true);
+  auto column = vectorMaker.flatVector<int64_t>(
+      size, [offset](auto row) { return row + offset; });
+
+  for (auto j = 0; j < FLAGS_num_batches; ++j) {
+    auto rowVectorBatch = input[j];
+    auto newNames = asRowType(rowVectorBatch->type())->names();
+    newNames.push_back(columnName);
+    auto newChildren = rowVectorBatch->children();
+    newChildren.push_back(column);
```

Comment:
Are we going to replace these frame bound columns with the original offset columns before sending the input to PrestoQueryRunner? (cc: @aditi-pandit)

Reply:
@kagamiori : So actually we will have to think about the constant case. If the range end is a constant and not a column we will need to maintain it in some context.

Reply:
Hi @aditi-pandit, I see. So generating SQL for k-range frames with constant bound is actually less straightforward than frames with column bound. Then, compared to parsing "preceding_5", I'd prefer the previous idea you proposed of passing a context along with the plan to PrestoQueryRunner. "5" can then be provided there.

Reply:
@kagamiori : Yes, we have a mini-design.

@minhancao, @pramodsatya : Lets sync about this and share a doc with Wei.